packages/js/currency/src/utils.tsx

```diff
@@ -406,3 +406,100 @@ export function getCurrencyData() {
 		},
 	};
 }
+
+/**
+ * Escape special characters for user input in regex.
+ *
+ * @param {string} string
+ * @return {string} string
+ */
+const escapeRegExp = ( string: string ) => {
+	return string.replace( /[.*+?^${}()|[\]\\]/g, '\\$&' );
+};
+
+/**
+ * Localises a number or numeric string for display, adding the appropriate thousands and decimal separators.
+ * For compatibility reasons, it returns the input if it's not a number or a string of numbers.
+ */
+export const localiseMonetaryValue = (
+	config: NumberConfig,
+	number: number | string | unknown
+) => {
+	if ( typeof number === 'number' ) {
+		return numberFormat( config, number );
+	}
+
+	if ( typeof number === 'string' ) {
+		const dot = escapeRegExp( config.decimalSeparator );
+		const comma = escapeRegExp( config.thousandSeparator );
+
+		// Regex to match strictly numbers with arbitrary thousands and decimal separators.
+		// Example: /^\s*(\d+|\d{1,3}(?:,\d{3})*)(?:\.\d+)?\s*$/ for default config.
+		const regex = new RegExp(
+			`^\\s*(\\d+|\\d{1,3}(?:${ comma }\\d{3})*)(?:${ dot }\\d+)?\\s*$`
+		);
+
+		return number.replace( regex, ( n ) => {
+			const parsed = parseNumber( config, n );
+			return numberFormat( config, parsed );
+		} );
+	}
+
+	return number;
+};
+
+export const unformatLocalisedMonetaryValue = (
+	config: NumberConfig,
+	inputNumber: number | string | unknown
+) => {
+	if ( ! inputNumber ) {
+		throw new Error( 'Input value is undefined' );
+	}
+
+	if ( Number.isFinite( inputNumber ) ) {
+		return inputNumber;
+	}
+
+	if ( typeof inputNumber !== 'string' ) {
+		throw new Error( 'Input value is not a number or a numeric string' );
+	}
+
+	// Brackets signal a formula. Avoid unformatting these values.
+	if ( inputNumber.includes( '[' ) && inputNumber.includes( ']' ) ) {
+		throw new Error( 'Input value contains formula' );
+	}
+
+	// Check if the string contains any non-numeric characters except allowed separators and whitespace
+	const allowedChars = new RegExp(
+		`^\\s*[0-9${ escapeRegExp( config.thousandSeparator ) }${ escapeRegExp(
+			config.decimalSeparator
+		) }]+\\s*$`
+	);
+
+	if ( ! allowedChars.test( inputNumber ) ) {
+		throw new Error(
+			'Input value contains non-numeric characters and is not a formula'
+		);
+	}
+
+	if (
+		// check that there is only 1 decimal separator and it is not to the left of
+		// the thousands separator if there is a thousands separator in the value
+		inputNumber.split( config.decimalSeparator ).length > 2 ||
+		( inputNumber.includes( config.thousandSeparator ) &&
+			inputNumber.includes( config.decimalSeparator ) &&
+			inputNumber.indexOf( config.decimalSeparator ) <=
+				inputNumber.indexOf( config.thousandSeparator ) )
+	) {
+		throw new Error( 'Invalid decimal separator' );
+	}
+
+	const unformattedValue = inputNumber
+		.replace(
+			new RegExp( escapeRegExp( config.thousandSeparator ), 'g' ),
+			''
+		)
+		.replace( config.decimalSeparator, '.' );
```

Comment:
This now creates a conflict - the backend (`src/Utilities/NumberUtil.php`) expects a localized number, and the front-end sends an unformatted value.

A way to test this is to set the thousand separator to ` ` and the decimal separator to `,`, then providing this as the price: `1 234,12`

Reply:
Yikes, we really can't unlocalize twice: once on client and again on server - Good catch and thanks for bringing it up. 

On the server, what if we check to see if the string is in fact a legit number with `is_numeric` before unlocalizing?

If we don't do that I see a couple options:

1. unlocalize on the client only.
2. unlocalize on the server only.
3. Pass a parameter to say whats being sent (localized/unlocalized) and proceed accordingly.
4. Go back to `<input type="number" ... />` before all this fanciness started.

Logically 2. makes sense, but with the system designed for numbers we'd need to check to make sure the API isn't used elsewhere in the app or by extensions. I think the `is_numeric` check is the way to go.
```

Reply:
_grumble grumble_ why is this so hard 🥲

How about - on the backend:

1. if it has anything other than a single decimal point, attempt to delocalise according to localisation rules
2. if it has a single decimal point && is_numeric, accept it as a number (because is_numeric is slightly looser, [1](https://www.php.net/manual/en/function.is-numeric.php), [2](https://www.php.net/manual/en/language.types.numeric-strings.php))
3. else: error

Reply:
Yippie, sounds good to me

Reply:
@waclawjacek does the [commit](https://github.com/woocommerce/woocommerce/pull/54181/commits/5eb5e149fdd1dace6658749dcfb3d4646d917e2c) I pushed look ok to you? I'd like to get this into a 9.7.0 CFE soon